new-components/enrichers/custom-annotation/internal/annotation/annotation.go

```diff
@@ -0,0 +1,108 @@
+package annotation
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log/slog"
+	"slices"
+
+	"github.com/smithy-security/pkg/env"
+
+	"github.com/smithy-security/smithy/sdk/component"
+	ocsf "github.com/smithy-security/smithy/sdk/gen/com/github/ocsf/ocsf_schema/v1"
+)
+
+type (
+	// Conf contains the annotation configuration.
+	Conf struct {
+		AnnotationName   string
+		AnnotationValues map[string]any
+	}
+
+	customAnnotator struct {
+		conf *Conf
+	}
+)
+
+// NewConf initialises the enricher configuration from environment variables.
+func NewConf() (*Conf, error) {
+	var (
+		cfg = &Conf{
+			AnnotationValues: make(map[string]any),
+		}
+		err error
+	)
+
+	cfg.AnnotationName, err = env.GetOrDefault(
+		"CUSTOM_ANNOTATION_NAME",
+		"",
+		env.WithDefaultOnError(true),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	annotationValues, err := env.GetOrDefault("CUSTOM_ANNOTATION_VALUES", "{}")
+	if err != nil {
+		return nil, err
+	}
+
+	if len(annotationValues) > 0 && annotationValues != "{}" {
+		if err := json.Unmarshal([]byte(annotationValues), &cfg.AnnotationValues); err != nil {
+			return nil, fmt.Errorf("invalid JSON values supplied '%s': %w", annotationValues, err)
+		}
+	}
+
+	return cfg, nil
+}
+
+// NewCustomAnnotator returns a new custom annotator.
+func NewCustomAnnotator(conf *Conf) (*customAnnotator, error) {
+	switch {
+	case conf == nil:
+		return nil, errors.New("nil annotation configuration")
+	case conf.AnnotationName == "":
+		return nil, errors.New("annotation name cannot be empty")
+	}
+	return &customAnnotator{
+		conf: conf,
+	}, nil
+}
+
+// Annotate adds annotated values to passed findings.
+func (ca *customAnnotator) Annotate(
+	ctx context.Context,
+	findings []*ocsf.VulnerabilityFinding,
+) ([]*ocsf.VulnerabilityFinding, error) {
+	var (
+		providerName = "custom-annotation-enricher"
+		newFindings  = slices.Clone(findings)
+		logger       = component.LoggerFromContext(ctx).
+				With(slog.Int("num_findings", len(findings))).
+				With(slog.String("provider_name", providerName)).
+				With(slog.String("annotation_name", ca.conf.AnnotationName)).
+				With(slog.Any("annotation_values", ca.conf.AnnotationValues))
+	)
+
+	logger.Debug("preparing to annotate findings...")
+
+	for idx := range newFindings {
+		b, err := json.Marshal(ca.conf.AnnotationValues)
```

Comment:
this unmarshaling should be done once in the NewConf instead of being repeated for every loop since the resulting values will be the same